### PR TITLE
Configure performance counters collection

### PR DESCRIPTION
**Pull request by lmolkova**
Fix issue https://github.com/Azure/azure-functions-host/issues/3852

Question: do we want to put it under the configuration so users who don't want it could turn it off?

### CODE REVIEW

**Comment by brettsam** [comment]
And yes, I think we should add on/off switches for a bunch of these modules/processors. I've added #2123 to track that.

**Comment by lmolkova** [comment]
It enables counters provided by Kudu - https://github.com/projectkudu/kudu/wiki/Perf-Counters-exposed-as-environment-variables (WEBSITE_COUNTERS_ALL). 

```json
{
  "aspNet": {
    "applicationRestarts": 0,
    "applicationsRunning": 0,
    "requestsDisconnected": 0,
    "requestExecutionTime": 0,
    "requestsRejected": 0,
    "requestsQueued": 0,
    "wpsRunning": 0,
    "wpsRestarts": 0,
    "requestWaitTime": 0,
    "requestsCurrent": 0,
    "globalAuditSuccess": 0,
    "globalAuditFail": 0,
    "globalEventsError": 0,
    "globalEventsHttpReqError": 0,
    "globalEventsHttpInfraError": 0,
    "requestsInNativeQueue": 0,
    "anonymousRequests": 0,
    "totalCacheEntries": 0,
    "totalCacheTurnoverRate": 0,
    "totalCacheHits": 0,
    "totalCacheMisses": 0,
    "totalCacheRatioBase": 0,
    "apiCacheEntries": 0,
    "apiCacheTurnoverRate": 0,
    "apiCacheHits": 0,
    "apiCacheMisses": 0,
    "apiCacheRatioBase": 0,
    "outputCacheEntries": 0,
    "outputCacheTurnoverRate": 0,
    "outputCacheHits": 0,
    "outputCacheMisses": 0,
    "outputCacheRatioBase": 0,
    "compilations": 0,
    "debuggingRequests": 0,
    "errorsPreProcessing": 0,
    "errorsCompiling": 0,
    "errorsDuringRequest": 0,
    "errorsUnhandled": 0,
    "errorsTotal": 0,
    "pipelines": 0,
    "requestBytesIn": 0,
    "requestBytesOut": 0,
    "requestsExecuting": 0,
    "requestsFailed": 0,
    "requestsNotFound": 0,
    "requestsNotAuthorized": 0,
    "requestsInApplicationQueue": 0,
    "requestsTimedOut": 0,
    "requestsSucceded": 0,
    "requestsTotal": 0,
    "sessionsActive": 0,
    "sessionsAbandoned": 0,
    "sessionsTimedOut": 0,
    "sessionsTotal": 0,
    "transactionsAborted": 0,
    "transactionsCommitted": 0,
    "transactionsPending": 0,
    "transactionsTotal": 0,
    "sessionStateServerConnections": 0,
    "sessionSqlServerConnections": 0,
    "eventsTotal": 0,
    "eventsApp": 0,
    "eventsError": 0,
    "eventsHttpReqError": 0,
    "eventsHttpInfraError": 0,
    "eventsWebReq": 0,
    "auditSuccess": 0,
    "auditFail": 0,
    "memberSuccess": 0,
    "memberFail": 0,
    "formsAuthSuccess": 0,
    "formsAuthFail": 0,
    "viewstateMacFail": 0,
    "appRequestExecTime": 0,
    "appRequestDisconnected": 0,
    "appRequestsRejected": 0,
    "appRequestWaitTime": 0,
    "cachePercentMachMemLimitUsed": 0,
    "cachePercentMachMemLimitUsedBase": 0,
    "cachePercentProcMemLimitUsed": 0,
    "cachePercentProcMemLimitUsedBase": 0,
    "cacheTotalTrims": 0,
    "cacheApiTrims": 0,
    "cacheOutputTrims": 0,
    "appCpuUsed": 0,
    "appCpuUsedBase": 0,
    "appMemoryUsed": 0,
    "requestBytesInWebsockets": 0,
    "requestBytesOutWebsockets": 0,
    "requestsExecutingWebsockets": 0,
    "requestsFailedWebsockets": 0,
    "requestsSucceededWebsockets": 0,
    "requestsTotalWebsockets": 0
  },
  "app": {
    "userTime": 65625000,
    "kernelTime": 38750000,
    "pageFaults": 175585,
    "processes": 2,
    "processLimit": 32,
    "threads": 37,
    "threadLimit": 512,
    "connections": 0,
    "connectionLimit": 300,
    "sections": 3,
    "sectionLimit": 256,
    "namedPipes": 2,
    "namedPipeLimit": 128,
    "readIoOperations": 2485,
    "writeIoOperations": 859,
    "otherIoOperations": 38155,
    "readIoBytes": 7715020,
    "writeIoBytes": 5850202,
    "otherIoBytes": 1899661,
    "privateBytes": 19206144,
    "handles": 2408,
    "contextSwitches": 365042,
    "remoteOpens": 3196,
    "remoteWrites": 47,
    "remoteWriteKBs": 952
  },
  "clr": {
    "bytesInAllHeaps": 0,
    "gcHandles": 0,
    "gen0Collections": 0,
    "gen1Collections": 0,
    "gen2Collections": 0,
    "inducedGC": 0,
    "pinnedObjects": 0,
    "committedBytes": 0,
    "reservedBytes": 0,
    "timeInGC": 0,
    "timeInGCBase": 0,
    "allocatedBytes": 0,
    "gen0HeapSize": 0,
    "gen1HeapSize": 0,
    "gen2HeapSize": 0,
    "largeObjectHeapSize": 0,
    "currentAssemblies": 0,
    "currentClassesLoaded": 0,
    "exceptionsThrown": 0,
    "appDomains": 0,
    "appDomainsUnloaded": 0
  }
}
```

**Comment by lmolkova** [comment]
> And yes, I think we should add on/off switches for a bunch of these modules/processors. I've added #2123 to track that.

@brettsam  I've tried to address it for perf counters in this way:

```csharp
services.TryAddSingleton<ITelemetryModule>(provider =>
{
    ApplicationInsightsLoggerOptions options = provider.GetService<IOptions<ApplicationInsightsLoggerOptions>>().Value;
    if (options.EnablePerformanceCountersCollection)
    {
        return new PerformanceCollectorModule();
    }

    return new NoopTelemetryModule();
});
```

I've tried with no luck to find better way so that based on options we won't add anything. If anyone knows how to do it - please let me know.

**Comment by brettsam** [comment]
@lmolkova can you squash these changes down to a single commit?

File: src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/NoopTelemetryModule.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    /// <summary>
+    /// Noop telemetry module. Used when customers disables certain modules.
+    /// </summary>
+    internal class NoopTelemetryModule : ITelemetryModule
+    {
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+        }
+    }
+}
```
Comment (brettsam) [inline src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/NoopTelemetryModule.cs]: Can you port the `NullTelemetryModule` change from https://github.com/Azure/azure-webjobs-sdk/pull/2073 over to here? We'll get this PR in first...

Comment (lmolkova) [reply src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/NoopTelemetryModule.cs]: Sure, I was waiting for #2073 to get merged so I can rebase and avoid conflicts. Do you want to merge this one first?

Comment (brettsam) [reply src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/NoopTelemetryModule.cs]: Yeah, let's do this one first since it's less impactful.

Comment (brettsam) [reply src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/NoopTelemetryModule.cs]: And actually... not to make this too big, but can you just move ApplicationInsights forward to the latest nugets here as well? That way if #2073 slips for any reason, we'll still get that change into this release next week.

**Comment by brettsam** [comment]
Looks good to me! Can you squash this down and once tests all pass, I will merge?

**Pull request merged**